toolchain/check/member_access.cpp

```diff
@@ -570,29 +571,32 @@ auto PerformTupleIndex(Context& context, SemIR::LocId loc_id,
       context.GetBuiltinType(SemIR::BuiltinInstKind::IntType));
   auto index_const_id = context.constant_values().Get(index_inst_id);
   if (index_const_id == SemIR::ConstantId::Error) {
-    index_inst_id = SemIR::InstId::BuiltinError;
+    return SemIR::InstId::BuiltinError;
   } else if (!index_const_id.is_template()) {
     // TODO: Decide what to do if the index is a symbolic constant.
     CARBON_DIAGNOSTIC(TupleIndexNotConstant, Error,
                       "Tuple index must be a constant.");
     context.emitter().Emit(loc_id, TupleIndexNotConstant);
-    index_inst_id = SemIR::InstId::BuiltinError;
-  } else {
-    auto index_literal = context.insts().GetAs<SemIR::IntLiteral>(
```

Comment:
This change is necessary since without early returning a `SemIR::InstId::BuiltinError` I hit this https://github.com/carbon-language/carbon-lang/blob/06344aeb7cd22c3a8b32cc7b48d49a57f8bcbafa/toolchain/check/eval.cpp#L468 on this test https://github.com/carbon-language/carbon-lang/blob/06344aeb7cd22c3a8b32cc7b48d49a57f8bcbafa/toolchain/check/testdata/var/no_prelude/fail_init_type_mismatch.carbon#L15
Which is a weird bug, since it only happens when `--no-prelude-import` is passed, and seems to be okay without.

Reply:
I think the root issue here is that all the existing code that works with `ElementIndex` values isn't prepared to handle invalid values (since until now they weren't possible). So without this change, this function was emitting an invalid `ElementIndex`, and then some code somewhere tried to use it as if it were a valid index. And I think the fix you have here is the right one: make sure an invalid `ElementIndex` never leaves this function.